cub/cub/device/dispatch/dispatch_transform.cuh

```diff
@@ -0,0 +1,1161 @@
+{
+  extern __shared__ char smem[]; // this should be __attribute((aligned(memcpy_async_alignment))), but then it clashes
+                                 // with the ublkcp kernel, which sets a higher alignment, since they are both called
+                                 // from the same kernel entry point (albeit one is always discarded). However, SMEM is
+                                 // 16-byte aligned by default.
+
+  constexpr int block_dim  = MemcpyAsyncPolicy::BLOCK_THREADS;
+  const Offset tile_stride = block_dim * num_elem_per_thread;
+  const Offset offset      = static_cast<Offset>(blockIdx.x) * tile_stride;
+  const int tile_size      = ::cuda::std::min(num_items - offset, tile_stride);
+
+  auto group           = cooperative_groups::this_thread_block();
+  int smem_offset      = 0;
+  const auto smem_ptrs = ::cuda::std::tuple<const InTs*...>{
+    copy_and_return_smem_dst(group, tile_size, smem, smem_offset, offset, aligned_ptrs)...};
+  cooperative_groups::wait(group);
+  (void) smem_ptrs; // suppress unused warning for MSVC
+  (void) &smem_offset; // MSVC needs extra strong unused warning supression
+
+  {
+    // move the whole index and iterator to the block/thread index, to reduce arithmetic in the loops below
+    num_items -= offset;
+    out += offset;
+  }
+
+#pragma unroll 1
+  for (int i = 0; i < num_elem_per_thread; ++i)
+  {
+    const int idx = i * block_dim + threadIdx.x;
+    if (idx < num_items)
+    {
+      out[idx] = poor_apply(
+        [&](const InTs*... smem_base_ptrs) {
+          return f(smem_base_ptrs[idx]...);
+        },
+        smem_ptrs);
+    }
+  }
+}
+
+constexpr int ublkcp_alignment     = 128;
+constexpr int ublkcp_size_multiple = 16;
+
+#ifdef _CUB_HAS_TRANSFORM_UBLKCP
+// TODO(bgruber): inline this as lambda in C++14
+template <typename Offset, typename T>
+_CCCL_DEVICE void copy_ptr_set(
+  ::cuda::std::uint64_t& bar,
+  int tile_size,
+  int tile_stride,
+  char* smem,
+  int& smem_offset,
+  ::cuda::std::uint32_t& total_copied,
+  Offset global_offset,
+  const aligned_base_ptr<T>& aligned_ptr)
+{
+#  if CUB_PTX_ARCH >= 900
+  namespace ptx = ::cuda::ptx;
+  // Copy a bit more than tile_size, to cover for base_ptr starting earlier than ptr
+  const uint32_t num_bytes = round_up_to_po2_multiple(
+    aligned_ptr.offset + static_cast<uint32_t>(sizeof(T)) * tile_size, static_cast<uint32_t>(ublkcp_size_multiple));
+  ptx::cp_async_bulk(
+    ptx::space_cluster,
+    ptx::space_global,
+    smem + smem_offset,
+    aligned_ptr.ptr + global_offset, // Use 128-byte aligned base_ptr here
+    num_bytes,
+    &bar);
+  smem_offset += static_cast<int>(sizeof(T)) * tile_stride + 128;
+  total_copied += num_bytes;
+#  endif // CUB_PTX_ARCH >= 900
+};
+
+// TODO(bgruber): inline this as lambda in C++14
+template <typename T>
+_CCCL_DEVICE _CCCL_FORCEINLINE const T&
+fetch_operand(int tile_stride, const char* smem, int& smem_offset, int smem_idx, const aligned_base_ptr<T>& aligned_ptr)
```

Comment:
Question: Can the alignment requirements of `T` ever exceed 128-bytes? (i.e. is there a max alignment specified by the C++ spec?)

In that case, it may be that the reinterpret_cast does not satisfy the alignment requirements of `T`.

Suggestion: move all types with too large alignment requirements to the fallback kernel.

Reply:
A type with alignment larger than `std::max_align_t` (16 bytes on my system) is called an overaligned type. The amount of overalignment supported is implementation defined. nvcc/g++ has no troubles with huge numbers here. So yes, we can have types with alignment larger than 128 bytes.